OpenTelemetry Part 3: v0.6.0 gems and VMPooler (#47)
genebean/jekyll-beanbag: _posts/2020-09-22-otel-part-3-v060.md

The combination of Gitter and the SIG meetings on Zoom have been incredibly  helpful.

## Tickets pre v0.6.0

Before v0.6.0 came out I opened [Unable to add tags to spans #312](https://github.com/open-telemetry/opentelemetry-ruby/issues/312). The discussion on that ticket was very educational. It also lead to [Export resources from Jaeger #348](https://github.com/open-telemetry/opentelemetry-ruby/pull/348) being done and included in the 0.6.0 release. One of the key reasons this is important is that it allows for setting both basic and custom attributes (tags) that prior to 0.6.0 had to be tacked on by the external Jaeger agent.

> One fo the key tickets I was watching was [OTLP exporter #277](https://github.com/open-telemetry/opentelemetry-ruby/pull/277) but [protocolbuffers/protobuf#1594](https://github.com/protocolbuffers/protobuf/issues/1594) threw a big wrench into it being a path forward for me: right now is that right now the gRPC used by OTLP doesn't support JRuby... most of the apps I am working on run on JRuby. Fortunately, [#231](https://github.com/open-telemetry/opentelemetry-ruby/issues/231) was already scheduled to be in the 0.6.0 release. That issues called for implementing Binary Thrift over HTTP as a transport for the Jaeger exporter.

Author: ```suggestion
One of the key tickets I was watching was [OTLP exporter #277](https://github.com/open-telemetry/opentelemetry-ruby/pull/277) but [protocolbuffers/protobuf#1594](https://github.com/protocolbuffers/protobuf/issues/1594) threw a big wrench into it being a path forward for me: the problem is that right now the gRPC used by OTLP doesn't support JRuby... most of the apps I am working on run on JRuby. Fortunately, [#231](https://github.com/open-telemetry/opentelemetry-ruby/issues/231) was already scheduled to be in the 0.6.0 release. That issues called for implementing Binary Thrift over HTTP as a transport for the Jaeger exporter.
```

> The reason that, initially, I was watching #277 and then watched #231 is that either of those being done would mean I no longer had to have a local Jaeger agent; I could, instead, send traces directly from the application via TCP-based protocol directly to the OpenTelemetry collector.

Author: ```suggestion
The reason that, initially, I was watching #277 and then watched #231 is that either of those being done would mean I no longer had to have a local Jaeger agent; I could, instead, send traces directly from the application via a TCP-based protocol directly to the OpenTelemetry collector.
```

## Upgrading to v0.6.0

Once the new release came out, it was time to start upgrading so that attributes could be set in code and Jaeger agents could be ditched. It also meant that I could start working in earnest on VMPooler since there was no longer a dependency on the agent. Both sets of work started in parallel at this point... which, in hind sight, might not have been my best plan. Read on to see what I mean.

## VMPooler part 1: working locally

> Things started out pretty simple When I started working on adding the OTel instrumentation to VMPooler. That is, until I ran `docker-compose up` and watched something about the new tracing code cause the app to crash with only this error as a clue: `E, [2020-09-12T00:36:44.445784 #1] ERROR -- : unexpected error in Jaeger::CollectorExporter#export - Not enough bytes remain in buffer`. I started digging and found some Jaeger docs for how to [Increase in-memory queue size](https://www.jaegertracing.io/docs/1.18/performance-tuning/#increase-in-memory-queue-size). My initial impression after reading those docs was that I needed the Jaeger exporter to either flush faster or have a bigger buffer. I couldn't find a way to do that but I did remember seeing that there was an alternative to the `SimpleSpanProcessor` called `BatchSpanProcessor`. Sadly, there were not any docs saying what span processor I should use or how to use each one. Fortunately I didn't give up and poked around in the repository on GitHub and discovered enough info to try it out by reading the comments in [batch_span_processor.rb](https://github.com/open-telemetry/opentelemetry-ruby/blob/master/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb). Though I didn't exactly understand why, I did find that swapping out the span processor fixed my issue.

Author: ```suggestion
Things started out pretty simple When I started working on adding the OTel instrumentation to VMPooler. That is, until I ran `docker-compose up` and watched something about the new tracing code cause the app to crash with only this error as a clue:

> E, [2020-09-12T00:36:44.445784 #1] ERROR -- : unexpected error in Jaeger::CollectorExporter#export - Not enough bytes remain in buffer

I started digging and found some Jaeger docs for how to [Increase in-memory queue size](https://www.jaegertracing.io/docs/1.18/performance-tuning/#increase-in-memory-queue-size). My initial impression after reading those docs was that I needed the Jaeger exporter to either flush faster or have a bigger buffer. I couldn't find a way to do that but I did remember seeing that there was an alternative to the `SimpleSpanProcessor` called `BatchSpanProcessor`. Sadly, there were not any docs saying what span processor I should use or how to use each one. Fortunately I didn't give up and poked around in the repository on GitHub and discovered enough info to try it out by reading the comments in [batch_span_processor.rb](https://github.com/open-telemetry/opentelemetry-ruby/blob/master/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb). Though I didn't exactly understand why, I did find that swapping out the span processor fixed my issue.
```

    span_processor = OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor.new(
      exporter: OpenTelemetry::SDK::Trace::Export::NoopSpanExporter.new
    )
  else
    puts "Exporting of traces will be done over HTTP in binary Thrift format to #{tracing_jaeger_host}"
    span_processor = OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor.new(
      exporter: OpenTelemetry::Exporter::Jaeger::CollectorExporter.new(
        endpoint: tracing_jaeger_host
      )
    )
  end

  OpenTelemetry::SDK.configure do |c|
    c.use 'OpenTelemetry::Instrumentation::Sinatra'
    c.use 'OpenTelemetry::Instrumentation::ConcurrentRuby'
    c.use 'OpenTelemetry::Instrumentation::Redis'

    c.add_span_processor(span_processor)

    c.resource = OpenTelemetry::Resource::Detectors::AutoDetector.detect
    c.resource = OpenTelemetry::SDK::Resources::Resource.create(
      {
        OpenTelemetry::SDK::Resources::Constants::SERVICE_RESOURCE[:name] => service_name,
        OpenTelemetry::SDK::Resources::Constants::SERVICE_RESOURCE[:version] => version
      }
    )
  end
end
```

The `self.configure_tracing` method isn't quite as complex as it may look. All that code breaks down to this:

> 1. set a variable for the service name to be used. This is needed because VMPooler can run as two independant servies (api and manager) or as an all-in-one.

Author: ```suggestion
1. set a variable for the service name to be used. This is needed because VMPooler can run as two independent services (api and manager) or as an all-in-one.
```

2. update the service name to include the defined prefix which usually maps to a instance's name such as prod, stage, or test.
3. set the span exporter based on if tracing is enabled or not. When it's disabled the `NoopSpanExporter` get's used so that no trace data is emmitted. When tracing is enabled the endpoint to which to send the data to is also configured.
4. run OTel's configurator and tell it to:
   1. enable automated instrumentation for Sinatra, ConcurrentRuby, and Redis
   2. use the selected span processor
   3. enable automatic resource detection
   4. set `service.name` and `service.version` (note that this is using the verbose method - I'll update it after [#426](https://github.com/open-telemetry/opentelemetry-ruby/pull/426) is merged into the OTel gems)

### bin/vmpooler

> VMPooler is actually run by calling `bin/vmpooler` so that is where the final bit go. First, I needed to bring in a few new settings and make a few new local variables. I did that by added these lines to the file:

Author: ```suggestion
VMPooler is actually run by calling `bin/vmpooler` so that is where the final bit go. First, I needed to bring in a few new settings and make a few new local variables. I did that by adding these lines to the file:
```

```ruby
require 'vmpooler/version'

prefix = config[:config]['prefix']
tracing_enabled = config[:tracing]['enabled']
tracing_jaeger_host = config[:tracing]['jaeger_host']

version = Vmpooler::VERSION

startup_args = ARGV
```

With those in place I was able to call the helper method that was added to `lib/vmpooler` by adding this line:

```ruby
Vmpooler.configure_tracing(startup_args, prefix, tracing_enabled, tracing_jaeger_host, version)
```

> And that's all the ruby code that was needed. Beyond that, our Dockerfiles did need the small adjustement of adding the installation of `make`.

Author: ```suggestion
And that's all the ruby code that was needed. Beyond that, our Dockerfiles did need the small adjustment of adding the installation of `make` so that all the gems would properly install.
```

## What's next?

> My Sinatra journey is progressing nicely but there is still more to do. Next up is adding some manual instrumentation to VMPooler and ABS and upgrading to 0.7.0 as soon as it comes out so that I can get the chagnes talked about above.

Author: ```suggestion
My Sinatra journey is progressing nicely but there is still more to do. Next up is adding some manual instrumentation to VMPooler and ABS and upgrading to 0.7.0 as soon as it comes out so that I can get the changes talked about above.
```

With regards to the manual instrumentation part, I have started working on that in [Add additional data to spans in api/v1.rb #400](https://github.com/puppetlabs/vmpooler/pull/400) but want to do some manual testing before asking for it to be merged. My learnings there will directly influence how I move forward on a similar PR for ABS.

> ## Bonus: putting the OTel's Java agent to use

Author: ```suggestion
## Bonus: putting the OTel Java agent to use
```
